### ISSUE TITLE
Intermittent Error: Unable to locate Storage Account when RAGRS/RAZGRS account kind is used

### ISSUE THREAD

**Issue by andrey-dubnik** [hashicorp/terraform-provider-azurerm#15048]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

2.29.0

### Affected Resource(s)

azurerm_storage_account intermittently produces an error on plan for the already created resource `Error: Unable to locate Storage Account`

### Terraform Configuration Files

```
resource "azurerm_resource_group" "example" {
  name     = "example-resources"
  location = "West Europe"
}

resource "azurerm_storage_account" "example" {
  name                     = "storageaccountname"
  resource_group_name      = azurerm_resource_group.example.name
  location                 = azurerm_resource_group.example.location
  account_tier             = "Standard"
  account_replication_type = "RAZGRS"
}
```

### Expected Behaviour

Should be no error for the RA and non RA accounts

### Actual Behaviour

intermittently produces an error on plan for the already created resource `Error: Unable to locate Storage Account`

### Steps to Reproduce

1. `terraform apply`
2. `terraform plan`

### Important Factoids

Only affecting RA storage accounts

**Comment by magodo** [comment]
@andrey-moor I've applied the configuration as you provided above (a typo in the `account_replication_type`, which should be `RAGZRS` rather than `RAZGRS`). Then I started a loop to run `terraform plan` until it returns error. With a bunch of iterations, the issue doesn't appear.
So would you please follow the [terraform debug guide](https://www.terraform.io/internals/debugging) and provide the debug log here? Mostly that is due to some API issues, while we will need the "request id" and other context info for further contacting the Azure supports.

**Comment by andrey-dubnik** [comment]
Forgot to mention the issue was reproducible only via GH actions and was intermittent. It may be down to the Azure portal replication etc. as GH hosts runners in multiple location and 1st available runner is selected for the pipeline.

East US (eastus)
East US 2 (eastus2)
West US 2 (westus2)
Central US (centralus)
South Central US (southcentralus)

Similar behaviour is affecting the KeyVault but with much less frequency

Let me see if I can reproduce the issue and capture the debug data via the standalone pipeline, we have switched to ZGRS accounts to unblock the delivery so original config is not available.

**Comment by varnav** [comment]
I'm seeing this:

> Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="StorageAccountNotFound" Message="The storage account testsa was not found."

with 2.93.1

**Comment by andrey-dubnik** [comment]
@magodo sorry took me a while, I have captured the trace debug of the problem. Cant see anything apparent which might point to the issue... Here is the trace log https://gist.github.com/andrey-dubnik/575429afc1f057f04a2d7468ba5a7dae

**Comment by andrey-dubnik** [comment]
suspect it is related to the call which comes back empty - https://docs.microsoft.com/en-us/rest/api/storagerp/storage-accounts/list

```
2022-02-03T16:14:30.610Z [DEBUG] provider.terraform-provider-azurerm_v2.94.0_x5: AzureRM Response for https://management.azure.com/subscriptions/264f194d-71cc-41d9-9bf4-c4f5456285e8/providers/Microsoft.Storage/storageAccounts?api-version=2021-04-01: 
HTTP/2.0 200 OK
Cache-Control: no-cache
Content-Type: application/json; charset=utf-8
Date: Thu, 03 Feb 2022 16:14:30 GMT
Expires: -1
Pragma: no-cache
Strict-Transport-Security: max-age=31536000; includeSubDomains
Vary: Accept-Encoding
X-Content-Type-Options: nosniff
X-Ms-Correlation-Request-Id: fc3a2733-bb7a-9fdc-09e6-d9510d90bf10
X-Ms-Request-Id: 612cc49f-6921-4522-9a25-c7eaa4e7b3b2
X-Ms-Routing-Request-Id: EASTUS2:20220203T161430Z:612cc49f-6921-4522-9a25-c7eaa4e7b3b2

{"value":[]}: timestamp=2022-02-03T16:14:30.609Z
```

**Comment by andrey-dubnik** [comment]
Seem the workaround is to create a TAG on the storage account (via portal), after that the issue goes away and above API returns the value. More to it it started to return value for all the storage account in that sub which were having the issue...

**Comment by andrey-dubnik** [comment]
The workaround came from the #11059 as we've seen similar but not that permanent issue with the keyvault. The root cause is highly likely to be the same for both of those issues.

**Comment by magodo** [comment]
@andrey-dubnik Just to be sure, the `LIST` call for the SA on your test sub always return an empty list, even after waiting for a long while? Is there any other SA in that sub?

**Comment by andrey-dubnik** [comment]
This is correct, terraform was able to obtain the keys and all the data for the account but the list api returned blanc hence the account can't be found error. 

There was another account in the sub which was having an issue originally so in total there are 2 accounts in there. After tagging at least one the second account also appeared in the api call.

**Comment by roehrijn** [comment]
This issue is also causing sleepless nights on our side here. What I can say so far: most likely this issue is caused by some race condition hitting sorts of ARM API limits. We're experiencing this on a Terraform project with around 100 resources (1 storage account, 1 key vault, a lot of role assignments to them) and I'm not able to reproduce the issue in a smaller project. However, calling `terraform apply -parallelism=1` seems to help.

Furthermore: 
We're experiencing the very same behaviour also on Key Vaults (is switches randomly from Storage Account to Key Vault, whatever Terraform tries to access first). Here is a log example from that:

        2022-03-03T06:57:41.0852887Z 2022-03-03T06:57:40.565Z [INFO]  Starting apply for azurerm_key_vault_secret.githubtoken
        2022-03-03T06:57:41.0853464Z 2022-03-03T06:57:40.565Z [DEBUG] azurerm_key_vault_secret.githubtoken: applying the planned Update change
        2022-03-03T06:57:41.0854255Z 2022-03-03T06:57:40.567Z [INFO]  provider.terraform-provider-azurerm_v2.98.0_x5: preparing arguments for AzureRM KeyVault Secret update.: timestamp=2022-03-03T06:57:40.566Z
        2022-03-03T06:57:41.0854946Z 2022-03-03T06:57:40.567Z [DEBUG] provider.terraform-provider-azurerm_v2.98.0_x5: AzureRM Request: 
        2022-03-03T06:57:41.0855810Z GET /subscriptions/***/resources?%24filter=resourceType+eq+%27Microsoft.KeyVault%2Fvaults%27+and+name+eq+%27net-prod-839977%27&%24top=5&api-version=2020-06-01 HTTP/1.1
        2022-03-03T06:57:41.0856269Z Host: management.azure.com
        2022-03-03T06:57:41.0857164Z User-Agent: Go/go1.17.5 (amd64-linux) go-autorest/v14.2.1 Azure-SDK-For-Go/v61.4.0 resources/2020-06-01 HashiCorp Terraform/1.1.6 (+https://www.terraform.io) Terraform Plugin SDK/2.10.1 terraform-provider-azurerm/2.98.0 pid-222c6c49-1b0a-5959-a213-6608f9eb8820
        2022-03-03T06:57:41.0857908Z X-Ms-Correlation-Request-Id: b01b3074-24a4-9613-dd3f-c0a93338159e
        2022-03-03T06:57:41.0858375Z Accept-Encoding: gzip: timestamp=2022-03-03T06:57:40.566Z
        2022-03-03T06:57:41.0859118Z 2022-03-03T06:57:40.635Z [ERROR] vertex "azurerm_key_vault_secret.githubtoken" error: Unable to determine the Resource ID for the Key Vault at URL "https://net-prod-839977.vault.azure.net/"
        2022-03-03T06:57:41.0860401Z 2022-03-03T06:57:40.635Z [DEBUG] provider.terraform-provider-azurerm_v2.98.0_x5: AzureRM Response for https://management.azure.com/subscriptions/***/resources?%24filter=resourceType+eq+%27Microsoft.KeyVault%2Fvaults%27+and+name+eq+%27net-prod-839977%27&%24top=5&api-version=2020-06-01: 
        2022-03-03T06:57:41.0861012Z HTTP/2.0 200 OK
        2022-03-03T06:57:41.0861306Z Cache-Control: no-cache
        2022-03-03T06:57:41.0861671Z Content-Type: application/json; charset=utf-8
        2022-03-03T06:57:41.0861971Z Date: Thu, 03 Mar 2022 06:57:39 GMT
        2022-03-03T06:57:41.0862330Z Expires: -1
        2022-03-03T06:57:41.0862605Z Pragma: no-cache
        2022-03-03T06:57:41.0863019Z Strict-Transport-Security: max-age=31536000; includeSubDomains
        2022-03-03T06:57:41.0863386Z Vary: Accept-Encoding
        2022-03-03T06:57:41.0863725Z X-Content-Type-Options: nosniff
        2022-03-03T06:57:41.0864181Z X-Ms-Correlation-Request-Id: b01b3074-24a4-9613-dd3f-c0a93338159e
        2022-03-03T06:57:41.0864686Z X-Ms-Ratelimit-Remaining-Subscription-Reads: 11997
        2022-03-03T06:57:41.0865152Z X-Ms-Request-Id: f6f7065a-d272-43ec-9bf3-526a820189ac
        2022-03-03T06:57:41.0865654Z X-Ms-Routing-Request-Id: WESTUS3:20220303T065740Z:f6f7065a-d272-43ec-9bf3-526a820189ac
        2022-03-03T06:57:41.0865936Z 
        2022-03-03T06:57:41.0866144Z {"value":[]}: timestamp=2022-03-03T06:57:40.634Z

**Comment by mariussm** [comment]
@roehrijn , I have tried with parallelism 1 now, and it does not resolve any issue when it comes to storage accounts at least. Maybe that workaround only works for key vaults?

**Comment by roehrijn** [comment]
Hi @mariussm, it also works for storage accounts in my environment. However, as a wrote, this is unfortunately likely to be some sort of race condition in rate limiting. That's why I think parallelism=1 is not a 100% fix/workaround. Hope MS is going to address this soon.

**Comment by damoodamoo** [comment]
@roehrijn / @andrey-dubnik  - we've experienced this issue over the past month or so: We found that a call to the `/resources?%24filter=resourceType+eq+%27Microsoft.KeyVault%` endpoint provided different results depending on the region that serviced that ARM API call. Essentially if the `X-Ms-Routing-Request-Id` value was the same as the location of the resource, the call returned the key vault - but if not it sporadically would return an empty array. This had a downstream effect on the rest of the TF deployment as it treated the resource as missing. It would be interesting for you to check the `X-Ms-Routing-Request-Id` of your calls to see if there's a correlation between successful and failed ones.

We have an open case with the ARM API team, and so far they've confirmed that it's an issue with the ARM API cross-region Cache not being updated quick enough. 

In terms of fixes / workarounds - they're currently a bit limited:
- Recommendation from the API team is to use the Azure Resource Graph for list operations, which has a 1 minute SLO for replication. Potentially a pretty big change to the provider code.
- Wait for the replication issues to die down + backend to improve
- Find some way of manually targeting the ARM API at the location that contains the resource

cc: @stuartleeks

**Comment by andrey-dubnik** [comment]
Using tags as a workaround worked so far and the portal cache was replicated. There were no re-occurrences of the issue since tagging. 

Since this is a provider-api scope there is no way to influence it externally. If this is a replication lag and not a permanent issue then adding a retry logic would probably help in mitigating the issue as worst case it would be 1 min SLO in oppose to an error which is already good enough. 

If tagging permanently fixes the issue maybe the api team can use this in the replication fix...

**Comment by 292368** [comment]
Hello . 

I had also experience same  Intermittent  issue  i.e  Unable to locate Storage Account for GRS S account.
root cause of this issue is ARM cache sync issue.   Contact to Microsoft to resync the account .   

It working for me  port resync of ARM cache 

Thanks 
Tushar P

**Comment by paweltucholskiigt** [comment]
I got this issue also. 
`az storage account list`
was listing []
adding tag did fix the problem so indeed this looks like state cache problem.

**Comment by Duchynko** [comment]
We're experiencing this problem in GitHub actions too.

`Run hashicorp/setup-terraform@v1.3.2`

`provider registry.terraform.io/hashicorp/azurerm v3.10.0`

**Comment by AvtsVivek** [comment]
I am experiencing the same problem on my local machine. [Here is the repo](https://github.com/AvtsVivek/AzureWithTerraform/tree/main/iac/2400600-static-website). 

I run [these commands](https://github.com/AvtsVivek/AzureWithTerraform/blob/main/iac/2400600-static-website/commands.sh) and I encounter this specifically when I run the destroy commands.

```t
terraform plan -destroy -out main.destroy.tfplan

terraform apply main.destroy.tfplan
```

>│ Error: Unable to locate Storage Account "staticwebsiteprfpim"!
│

The following shows plan command.

![image](https://user-images.githubusercontent.com/41255281/177114062-88095fd7-ed7c-4d6d-9ff4-df0a476f5bdb.png)

I get that error when I run the command for the first time. When I re-run the same command the second time, things run fine.

Same is the case with apply command as well.

**Comment by paalders** [comment]
When using GH Actions in combination with AzureRM and account_replication_type LRS you have the same problem.
```
resource "azurerm_storage_account" "..." {
  name                     = local.storage_account_name
  resource_group_name      = azurerm_resource_group.saltstack.name
  location                 = azurerm_resource_group.saltstack.location
  account_tier             = "Standard"
  account_replication_type = "LRS"
}

```
![image](https://user-images.githubusercontent.com/8476263/177758472-cc310096-cf60-4467-a561-5ae003f54e44.png)

**Comment by ekristen** [comment]
Using azurerm 3.16.0, storage account type LRS, this happens fairly frequently.

**Comment by magodo** [comment]
@ekristen and @paalders could you please help confirm that the storage account list API (i.e. `/subscriptions/{subscriptionId}/providers/Microsoft.Storage/storageAccounts`) doesn't return the expected storage account, while the [ARG API](https://docs.microsoft.com/en-us/azure/governance/resource-graph/first-query-portal) does?

**Comment by AndrzejK-Atende** [comment]
Same problem in my case:

Terraform v.1.2.6
azurerm: v.3.17.0

It does not count if SA is LRS, od ZRS, randomly I get error: "Unable to locate Storage Account"

**Comment by MLKiiwy** [comment]
Also same issue for me on **github actions + terraform cloud**. Also **LRS** type for replication.
The storage account cannot be found.

I can see it on the azure dashboard and on my local machine the command line tool return the correct list.

Terraform v.1.2.6
azurerm: v.3.17.0

**Comment by magodo** [comment]
@MLKiiwy Just want to be sure, which command you were running that returns you the correct list? Is it the "resource list" or Azure Resource Graph query?

**Comment by as9333** [comment]
I am facing the same issue. with azurerm 3.21.0. Any workarounds ?

**Comment by avooca** [comment]
Randomly impacting LRS Storage accounts azurerm 3.5.0.  Quite annoying...

**Comment by rickvanc** [comment]
Got the same issue, just created a storage account through TF and 5 minutes later it can't find it.

**Comment by brad-lucas** [comment]
Still seeing this for LRS storage accounts as of terraform 1.3.7 + azurerm 3.37.0.

**Comment by ekristen** [comment]
This is still a problem. Right now I have a terraform that simply won't find a storage account on refresh even though it's clearly right there.

**Comment by avooca** [comment]
I opened a case with MS about this issue in Oct 2022 and was told that ARM v W39+ was supposed to solve this issue on the backend.  I mention this only to suggest others open a case via Azure portal to push the backend team while we wait for the provider authors do whatever they can.  Case to reference: 2210070040004240

This problem can be a show-stopper for days when it pops up....

________________________________

**Comment by brad-lucas** [comment]
Following up here, the information previously provided by @andrey-dubnik and @damoodamoo was spot on. Though I won't repeat what they already alluded to above, there is apparently an additional alternative that was suggested to me by Azure Support when working through my ticket (2304260010001874, in case anyone else wants a pointer to this reference).

> The Azure Resource Manager team has recently made significant improvements to the replication flows and mechanisms, including large-scale enhancements of Azure's replication architecture, which **should entirely address this issue for list calls at the resource group level**. The improvements for subscription-level list calls will be rolling out incrementally to tenants over the next few months. However, we still recommend that you use [Azure Resource Graph](https://docs.microsoft.com/azure/governance/resource-graph/overview) whenever possible for resource list calls. Azure Resource Graph offers a one-minute SLO and is designed to performantly handle list calls at scale.

When creating the storage account, it looks like the behavior within `resourceStorageAccountRead` checks out:
- a `ListKeys` call is made to determine that the account has been, using the provided resource group
    - calls `https://management.azure.com/subscriptions/<subscription ID>/resourceGroups/<resource group name>/providers/Microsoft.Storage/storageAccounts/<storage account name>/listKeys`
- eventually, a call to helper method `FindAccount` (`terraform-provider-azurerm/internal/services/storage/client/helpers.go`) is made, which makes a subscription-level call to the ARM API
    - calls `https://management.azure.com/subscriptions/<subscription ID>/providers/Microsoft.Storage/storageAccounts`

Instead, this code path could use the `ListByResourceGroup` method on the `AccountsClient` (imported from `github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2021-09-01/storage`).

Unfortunately, I don't have the cycles right now to help provide any deeper investigation, but I'd be happy to come back to do so in a few weeks if someone hasn't been able to add more evidence.

cc @magodo

**Comment by magodo** [comment]
@brad-lucas Thank you for the news! We shall probably wait for the subscription-level list mitigate the caching issue. The `FindAccount` can't use the `ListByResourceGroup` unfortunately as this method is used among multiple places in the provider, where many of them only has access to the storage account's name.

**Comment by brad-lucas** [comment]
@magodo, can I convince you that the resource group name could/should be used when it's available, creating a new fork in the code paths? List queries are always going to be eventually consistent, but consistency guarantees are much more important when creating a storage account.

This code path has intermittently, but consistently, broken our automated deployments for almost a year at this point, doesn't have any workaround from a pure automation standpoint, and would be the same for anyone else using similar automation.

**Comment by magodo** [comment]
@brad-lucas, we've made a bunch of tries to workaround the caching issue, see #11059. Note that the resouce group name is not always available, e.g. during a read for a `azurerm_storage_container`.

For your specific ask:

- Per my experience, retry won't be too much helpful here
- We've tried to use ARG instead of the listing API, see #13719

**Comment by ro6it** [comment]
I am also facing this issue intermittently but I am using storage account tier as LRS.

Azure provider - 3.69, tried with 3.78 also same error 
Terraform version - 1.5.5 and tried with 1.5.7

Error

`
│ Error: unable to locate Storage Account "iseappstoragefhq2"
│
│   with module.loadbalancer_dns.azurerm_storage_account.ise-app-storage,
│   on ../../module/loadbalancer_dns/functionapp.tf line 8, in resource "azurerm_storage_account" "ise-app-storage":
│    8: resource "azurerm_storage_account" "ise-app-storage" {
`


Debug log

`
ount=1 diagnostic_warning_count=0 tf_rpc=ReadResource timestamp=2023-10-27T13:39:56.639+0530
2023-10-27T13:39:56.646+0530 [ERROR] provider.terraform-provider-azurerm_v3.69.0_x5: Response contains error diagnostic: diagnostic_summary="unable to locate Storage Account "iseappstoragefhq2"" tf_resource_type=azurerm_storage_account @caller=github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/diag/diagnostics.go:55 diagnostic_severity=ERROR diagnostic_detail= tf_proto_version=5.3 tf_provider_addr=provider tf_req_id=0933de58-93db-0874-0dd6-a134016be01d tf_rpc=ReadResource @module=sdk.proto timestamp=2023-10-27T13:39:56.639+0530
2023-10-27T13:39:56.644+0530 [ERROR] vertex "module.loadbalancer_dns.azurerm_storage_account.ise-app-storage" error: unable to locate Storage Account "iseappstoragefhq2"
`

**Comment by tkekkone** [comment]
For me this kept happening when I deployed even a completely new set of resources to new rg. After deploying the storage accounts, the `plan` stage failed because it did not find the storage account that was just deployed. Also `az storage account list` showed empty results. When I did `az login` and `az account set --subscription...` again, it finds the storage account and plan works.

**Comment by Uqqasha** [comment]
Still facing this issue intermittently on Storage Account Standard_LRS StorageV2.

Terraform version: 1.7.4
Azure provider: 3.91.0

Error on terraform apply
` Error: retrieving queue properties for Storage Account (Subscription: "xxx" Resource Group Name: "xxx"
   Storage Account Name: "xxx"): queues.Client#GetServiceProperties: Failure responding to request: StatusCode=403 -- 
   Original Error: autorest/azure: Service returned an error. Status=403 Code="AuthenticationFailed" Message="Server failed 
   to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.\
               with azurerm_storage_account.st,
               on xxx line 17, in resource "azurerm_storage_account" "st":
               17: resource "azurerm_storage_account" "st" {
`

Error on terraform destroy
`╷
  │ Error: unable to locate "Storage Account (Subscription: \"xxxx\"\nResource Group Name: \"xxx\"\nStorage Account Name: 
  |            \"xxx\")"
  │  
  │   with azurerm_storage_account.st,
  │   on xxx line 17, in resource "azurerm_storage_account" "st":
  │   17: resource "azurerm_storage_account" "st" {
  │ 
`

**Comment by alexivanov-danone** [comment]
I got this issue today, while deploying new resources into a clean subscription.
Not sure what this is about, i was creating 158 new resources, out of that 3 were storage accounts.

**Comment by c4milo** [comment]
We are also running into this.

**Comment by hrda81** [comment]
Same issue for me on two storage accounts with an account replication type of "GRS"

Terraform version: 1.7.5
Azure provider: 3.92.0

**Comment by jobeehichoya** [comment]
same issue I just encountered.

**Comment by musabmirza-amperon** [comment]
Same !

**Comment by musabmirza-amperon** [comment]
Still facing with terraform `1.7.5` and azurerm `3.97.1`

**Comment by alxndr13** [comment]
got this issue as well:

TF version: 1.7.4
Azure provider: 3.97.1

**Comment by pabloopez** [comment]
Same here :(

**Comment by scorpion35** [comment]
Same here. 
```
Terraform v1.8.0

.\.terraform\providers\registry.terraform.io\hashicorp\azurerm\3.98.0\windows_386\terraform-provider-azurerm_v3.98.0_x5.exe
```

It works in one run, but stopped working majority of the time.

**Comment by scorpion35** [comment]
> Same here.
> 
> ```
> Terraform v1.8.0
> 
> .\.terraform\providers\registry.terraform.io\hashicorp\azurerm\3.98.0\windows_386\terraform-provider-azurerm_v3.98.0_x5.exe
> ```
> 
> It works in one run, but stopped working majority of the time.
> 
> ```
> resource "azurerm_storage_account" "storage-account" {
>   resource_group_name = "storage-rg"
>   location            = "eastus"
>   name = "stgacestus3839"
> 
>   account_tier             = "Standard"
>   account_replication_type = "LRS"
>   min_tls_version          = "TLS1_2"
> }
> ```
> 
> Powershell equivalent is working fine though
> 
> ```
> Get-AzStorageAccount -ResourceGroupName "storage-rg"
> 
> # this returns the storage account properly
> ```
> 
> Seems the unreleased version 3.99.0 has some fixes around storage account? https://github.com/hashicorp/terraform-provider-azurerm/blob/main/CHANGELOG.md#3990-unreleased
> 
> Anyone know when this could be released?

Womp womp. Looks like 3.99.0 was released last night, but it doesn't fix the issue. Still seeing 
```
Planning failed. Terraform encountered an error while generating this plan.

╷
│ Error: unable to locate "Storage Account (Subscription: \"*****\"\nResource Group Name: \"*****\"\nStorage Account Name: \"****\")"
```

```
terraform init -upgrade

# Upgraded to terraform-provider-azurerm_v3.99.0_x5.exe (from 3.98.0)
```

**Comment by jkroepke** [comment]
We got this error with 3.99 today.

But our issue was, that we had setup AzureAD Authentication for the Storage Account `default_to_oauth_authentication=true`, but we missed to set `storage_use_azuread` at provider level.

**Comment by c4milo** [comment]
I got this using `v3.102.0` today. It is really frustrating.

**Comment by scorpion35** [comment]
Lot of people are running into this issue, and for a long time. Anyone know how to summon the this repo's gods? 

Tried to take a look at the code base, but it's in Go and I am aware of this language yet

@manicminer sorry to ping you directly, saw your name on the latest closed PR's approver. Can you help out summon this repo's gods please

**Comment by ekristen** [comment]
So ultimately this is an Azure problem. Their API sucks. They decided it's better to be fast then accurate and have take eventually consistent to heart, it will be eventually consistent but that could be 30secs or 10 hours.

HOWEVER, I believe the terraform provider could be better!! It could stub state, I could not error on 404 not found given or try harder to wait knowing that the Azure API is terrible.

I'm a firm believer that unless it's a direct 409 conflict error, state should be stubbed with azure because of its predictable conflicting naming conventions. I also think that given its eventual consistency that the provide must try harder, wait longer, or stub state for a subsequent run and not error on things like this.

**Comment by gctrevino** [comment]
I encountered this issue a few days ago when upgrading AzureRM from 2.90 to 3.107, I have tried the tag workaround and it does nothing to my problem. I am testing the change in a sandbox environment, in the dev environment everything is working perfectly.

**Sandbox environment**
```
#Can do a fresh deploy, but if I try to update, it fails consistently with Error: unable to locate Storage Account
AzureDevOps Pipeline
Terraform 1.0.2
AzureRM 3.107.0
```

**Dev environment**
```
AzureDevOps Pipeline
Terraform 1.0.2
AzureRM 2.90.0
```

**storage block**
```
resource "azurerm_storage_account" "learning_materials_storage_account" {
  name                            = "sa${var.location_abb}lm${var.project_key}${var.env}"
  resource_group_name             = var.rg_name
  location                        = var.location
  account_tier                    = "Standard"
  account_replication_type        = "RAGRS"
  enable_https_traffic_only       = true
  min_tls_version                 = "TLS1_2"
  allow_nested_items_to_be_public = "true"

  tags = var.tags
}
```

**Comment by gctrevino** [comment]
I am not saying that this is the solution, but, I noticed that when I ran `az storage account list` without specifying the subscription, it returned an empty list `[]`, so this left me thinking that maybe is the same for azurerm, specially because I use 3 different subscriptions. 

So, I explicitly added the tenant and subscription ids to the setup of the azurerm and it seems to be working for me now. It would be good if someone else could make the test on their side to see if this consistently fixes the issue.

```
provider "azurerm" {
  tenant_id       = var.tenant_id
  subscription_id = var.subscription_id

  features {...}
}
```

**Comment by chapmanc** [comment]
Just to add a note we are seeing this with azurerm at 3.98.0. 

Can we get an update @tombuildsstuff as it's been a couple Months since an update :).

**Comment by manicminer** [comment]
In the next provider release, we'll be updating the SDK used for storage accounts to use [hashicorp/go-azure-sdk](https://github.com/hashicorp/go-azure-sdk). Once that's been released, and we've addressed any potential regressions, we can take another look at this issue.

**Comment by samuel-baena-stenn** [comment]
Any updates? Still facing this today

**Comment by GavBurke** [comment]
> I opened a case with MS about this issue in Oct 2022 and was told that ARM v W39+ was supposed to solve this issue on the backend.  I mention this only to suggest others open a case via Azure portal to push the backend team while we wait for the provider authors do whatever they can.  Case to reference: 2210070040004240
> 
> This problem can be a show-stopper for days when it pops up....
> […](#)

I have recent reference 2411080050001756, closed on 9th December (2024) as being "a by design scenario situation" and told by their support that it's for us to figure out a wait/retry function... surely it's for azurerm provider to do that for us?!?  😠

**Comment by ekristen** [comment]
Best advice, don't use azure. I know that's not possible for everyone, not even for me. 

Unfortunately this will likely never get fixed. The maintainers have stated as much on other issues that due to the crappy nature of the Azure API and the inconsistent nature in which they are implemented, they don't have the time or energy to track all the failure conditions to implement a fix.

Search my username for open and closed issues where I've argued for this, you'll find their responses. Personally I think that unless it's a 409 conflict error, the provider should always stub the state, even if it's just the ID, a future plan will see the ID and all the missing data and either IaC matches or it does and it'll take the correct steps.

**Comment by chapmanc** [comment]
> In the next provider release, we'll be updating the SDK used for storage accounts to use [hashicorp/go-azure-sdk](https://github.com/hashicorp/go-azure-sdk). Once that's been released, and we've addressed any potential regressions, we can take another look at this issue.

@manicminer just checking if this has been released? And if there could be an update on fixing this.

**Comment by magodo** [comment]
👏 I've submit a PR to follow https://github.com/hashicorp/terraform-provider-azurerm/issues/15048#issuecomment-1527867656 to alleviate the issues, mainly for `azurerm_storage_account`. While the LIST call is still there for a couple of other resource types.

**Comment by katbyte** [comment]
Thank you for the report, and apologies for the slow follow up here.

This issue is linked to #28617 (Storage: Eliminating listing storage account when its resource ID is available to use GET), which merged and shipped in v4.22.0 and appears to have fixed it, so we are closing this one out as resolved.

If you are still running into this on the current (v5.x) version of the provider, please let us know with a comment or open a new issue referencing this one and we will take another look.
